zip-frost.rst

```diff
@@ -0,0 +1,243 @@
+
+FROST is one of such threshold signatures protocols. (TODO: more about FROST)
+
+
+Requirements
+============
+
+- All signatures generated by following this ZIP MUST be verified successfully
+  as Sapling or Orchard spend authorization signatures.
+- (Must keep same security requirements from the spec?)
+- (Randomizer security requirements?)
+- The party that generates alpha should also be the party that has access to the witness for the proof.
+
+Non-requirements
+================
+
+- (Talk about how proof generation is not covered?)
+
+
+Specification
+=============
+
+Re-randomized FROST
+-------------------
+
+To add re-randomization to FROST, follow the specification [#FROST]_ with the
+following modifications:
+
+In Round Two, the Coordinator also generates a random scalar α by following the
+RedDSA.GenRandom procedure and sends it to each signer along with the message
+and the set of signing commitments.
+
+The `sign` function is changed to receive α and incorporate it into the
+computation of `sig_share`. It is specified as the following: ::
+
+  Inputs:
+  - identifier, Identifier i of the signer. Note identifier will never equal 0.
+  - sk_i, Signer secret key share, a Scalar.
+  - group_public_key, public key corresponding to the group signing key,
+    an Element in G.
+  - nonce_i, pair of Scalar values (hiding_nonce, binding_nonce) generated in
+    round one.
+  - msg, the message to be signed (sent by the Coordinator).
+  - commitment_list =
+      [(j, hiding_nonce_commitment_j, binding_nonce_commitment_j), ...], a
+    list of commitments issued in Round 1 by each signer and sent by the Coordinator.
+    Each element in the list indicates the signer identifier j and their two commitment
+    Element values (hiding_nonce_commitment_j, binding_nonce_commitment_j).s
+    This list MUST be sorted in ascending order by signer identifier.
+  - alpha, the randomizer Scalar.
+
+  Outputs: a Scalar value representing the signature share
+
+  def sign(identifier, sk_i, group_public_key, nonce_i, msg, commitment_list, alpha):
+    # Encode the commitment list
+    encoded_commitments = encode_group_commitment_list(commitment_list)
+
+    # Compute the binding factor
+    binding_factor = compute_binding_factor(encoded_commitments, msg)
+
+    # Compute the group commitment
+    group_commitment = compute_group_commitment(commitment_list, binding_factor)
+
+    # Compute Lagrange coefficient
+    participant_list = participants_from_commitment_list(commitment_list)
+    lambda_i = derive_lagrange_coefficient(identifier, participant_list)
+
+    # Compute the per-message challenge
+    challenge = compute_challenge(group_commitment, group_public_key, msg)
+
+    # Compute the signature share
+    (hiding_nonce, binding_nonce) = nonce_i
+    sig_share = hiding_nonce + (binding_nonce * binding_factor) + (lambda_i * (sk_i + alpha / len(commitment_list))  * challenge)
```

Comment:
Is this formula correct? `sig_share` would need to be `hiding_nonce + (binding_nonce * binding_factor) + (lambda_i * sk_i + alpha / len(commitment_list)) * challenge)` if the intention is for this signature to verify as `K + alphaG`, as I believe given the coordinator's calculation of `alphaG`.

Reply:
Yep it's wrong, I'm working on fixing it. Thanks for spotting it!

Reply:
Happy to. I was concerned I'd have to change my work significantly, yet seems like I won't.

While this is the model I took, as it creates a leaderless protocol, it is also possible for the coordinator to simply add alpha * challenge to the signature, no? It'd still be broadcasted though so everyone knows which key they're signing for (as the key is part of the challenge in general, though I'd have to double check RedDsa's policy here).

This would translate into a protocol without a coordinator by having every participant do so on their own. The only issue I had with this was regarding implementation. I found it beneficial to create constructions where the share signing function advances the state. If you have to sign for multiple shares, you cannot advance the state as such, assuming advancing the state once to X prevents advancing again to X. While you could sign for (sk_i + alpha) in the first place, offering a single sign, you then have to strip alpha * challenge out for the part communicated. It ends up very annoying to implement, so I do prefer the inversion based model present here. Regardless, I wanted to discuss the single-signer modification model as well, while I had the opportunity, and see what other peoples' thoughts were.

Reply:
Yeah in the version I'm working on, I'm making the coordinator add `alpha * challenge` as you said. I think it would still be leaderless since all participants now alpha so they can all add it to the final signature?

Sorry, what do you mean by state?

Reply:
Changed formula in cf19f23

Reply:
To preface, my priority with FROST is modularity and composability, not Zcash, so bit of a different focus. I also want to work on consistency in the ecosystem though, which is why it's fully usable in an IETF compliant fashion, and why I'd like to hear this can be used without issue under this ZIP as well.

I have a function, sign_share, in my trait which allows for an arbitrary algorithm to specify signing stages. The following is an excerpt implementing FROST over a ring signature scheme.

```
  fn sign_share(
    &mut self,
    view: &FrostView<Ed25519>,
    nonce_sum: dfg::EdwardsPoint,
    b: dfg::Scalar,
    nonce: dfg::Scalar,
    msg: &[u8]
  ) -> dfg::Scalar {
    // Apply the binding factor to the H variant of the nonce
    self.AH.0 += self.AH.1 * b;

    // Use the transcript to get a seeded random number generator
    // The transcript contains private data, preventing passive adversaries from recreating this
    // process even if they have access to commitments (specifically, the ring index being signed
    // for, along with the mask which should not only require knowing the shared keys yet also the
    // input commitment masks)
    let mut rng = ChaCha12Rng::from_seed(self.transcript.rng_seed(b"decoy_responses"));

    self.msg = Some(msg.try_into().expect("CLSAG message should be 32-bytes"));

    #[allow(non_snake_case)]
    let (clsag, pseudo_out, p, c) = Clsag::sign_core(
      &mut rng,
      &self.image,
      &self.input(),
      self.mask(),
      &self.msg.as_ref().unwrap(),
      nonce_sum.0,
      self.AH.0.0
    );
    self.interim = Some(Interim { p, c, clsag, pseudo_out });

    let share = dfg::Scalar(nonce.0 - (p * view.secret_share().0));

    share
  }
```

Compared to EdDSA/moden Schnorr:

```
  fn sign_share(
    &mut self,
    params: &FrostView<C>,
    nonce_sum: C::G,
    _: C::F,
    nonce: C::F,
    msg: &[u8],
  ) -> C::F {
    let c = H::hram(&nonce_sum, &params.group_key(), msg);
    self.c = Some(c);
    schnorr::sign::<C>(params.secret_share(), nonce, c).s
  }
```

There are two relevant distinctions.
1) The former, which expects to only be called once, performs a state mutation which is invalid if applied multiple times. The latter does not.
2) The former is `nonce.0 - (p * view.secret_share().0`. The latter resolves to `nonce + (private_key * challenge)`. For some reason, the classical Schnorr subtraction was used when calculating the solution.

Accordingly, this trait must be made contextual of signing involving an additive offset in order to properly do so. That is not the case with the linearly applied offset (nor an offset which only one party applies, which is technically with leader, yet given a proper protocol for generating the randomizer, ends up meaning one person does one additional scalar addition other parties do not, which is not of significance besides the lack of ability to claim "leaderless").